Source/Csla.Data.EF7.Shared/DbContextManager.cs

```diff
@@ -0,0 +1,170 @@
+//-----------------------------------------------------------------------
+#if !MONO
+using System;
+using Microsoft.Data.Entity;
+
+namespace Csla.Data.EF7
+{
+  /// <summary>
+  /// Provides an automated way to reuse 
+  /// Entity Framework DbContext objects 
+  /// within the context
+  /// of a single data portal operation.
+  /// </summary>
+  /// <typeparam name="C">
+  /// Type of database 
+  /// DbContext object to use.
+  /// </typeparam>
+  /// <remarks>
+  /// This type stores the object context object 
+  /// in <see cref="Csla.ApplicationContext.LocalContext" />
+  /// and uses reference counting through
+  /// <see cref="IDisposable" /> to keep the data context object
+  /// open for reuse by child objects, and to automatically
+  /// dispose the object when the last consumer
+  /// has called Dispose."
+  /// </remarks>
+  public class DbContextManager<C> : IDisposable where C : DbContext
+  {
+    private static object _lock = new object();
+    private C _context;
+    private string _label;
+    private string ContextLabel { get; set; }
+
+    /// <summary>
+    /// Gets the ObjectContextManager object for the     /// specified database.
+    /// </summary>
+
+    public static DbContextManager<C> GetManager()
+    {
+      return GetManager(string.Empty);
+    }
+
+    /// <summary>
+    /// Gets the ObjectContextManager object for the 
+    /// specified database.
+    /// </summary>
+    /// <param name="database">Database name as shown in the config file.</param>
+    public static DbContextManager<C> GetManager(string database)
+    {
+      return GetManager(database, "default");
+    }
+   
+    /// <summary>
+    /// Gets the ObjectContextManager object for the 
+    /// specified database.
+    /// </summary>
+    /// <param name="database">
+    /// The database name or connection string.
+    /// </param>
+    /// <param name="label">Label for this context.</param>    
+    /// <returns>ContextManager object for the name.</returns>
+    public static DbContextManager<C> GetManager(string database, string label)
+    {
+      lock (_lock)
+      {
+        var contextLabel = GetContextName(database, label);
+        DbContextManager<C> mgr = null;
+        if (ApplicationContext.LocalContext.Contains(contextLabel))
+        {
+          mgr = (DbContextManager<C>)(ApplicationContext.LocalContext[contextLabel]);
+        }
+        else
+        {
+          mgr = new DbContextManager<C>(database, label);
+          mgr.ContextLabel = contextLabel;
+          ApplicationContext.LocalContext[contextLabel] = mgr;
+        }
+        mgr.AddRef();
+        return mgr;
+      }
+    }      
+
+    private DbContextManager(string database, string label)
+    {
+      _label = label;                
+      if (string.IsNullOrEmpty(database))
+      {
+        _context = (C) (Activator.CreateInstance(typeof (C)));
+      }
+      else
+      {
+        _context = (C)(Activator.CreateInstance(typeof(C), database));
```

Comment:
I specifically meant this call will fail on line 98:

```
(C)(Activator.CreateInstance(typeof(C), database));
```

There is no such constructor overload on `DbContext` that takes a database / connection string name in EF7.

Line 94 above works because it uses the default constructor for `DbContext` - and the unit test I added in this PR exercises this and is passing.

So basically this PR works as long as you don't want to use: `DbContextManager<T>.GetManager("somedbname")

I can add a failing unit test for that if you like..

I have an idea on how to solve this, but it would be a bit of a change to the "usual" API that you'd expect with `DbContextManager`

Basically I'm thinking you'd have to use it more like this (forgive if this doesn't quite compile, written directly in this chat and GitHub sadly doesn't have intellisense..)

```
DbContextManager<MyDbContext>.GetManager( (options) = >  {
 options.UseSqlServer("some connection string here") }
);
```

So my current idea is that DbContextManager would accept a callback, and prior to creating a new instance of DbContext, it would invoke the callback and allow the consumer to set the options.

In the example the `UseSqlServer` method is an extension method that is available to those that use the SqlServer relational provider for EF - but equally people could use whatever EF provider they liked and in which case they would have different extension methods to call (Postgres etc) because thats the way EF7 works :)

I'll submit something on this basis, and add a unit test around it.